### PR TITLE
[SYCLomatic][part2] Implement cmake script migration framework

### PR DESCRIPTION
**Pull request by tomflinda**
1. Add special migration process for cmake_minimum_required
2. Add lit test to verify migration of cmake_minimum_required

### CODE REVIEW

**Review by zhiweij1 — APPROVED** [review]

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/test/dpct/cmake_migation/case_009/rules.yaml
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_009/rules.yaml:1]: Please remove the empty file.

File: clang/test/dpct/cmake_migation/case_009/expected.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.24)
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_009/expected.txt:1]: could you try to merge the expected.txt and input.cmake,  to make it easy to check the migration result.

File: clang/lib/DPCT/SaveNewFiles.cpp [hunk 2]
```diff
@@ -395,6 +396,11 @@ void applyPatternRewriterToCmakeScriptFile(const std::string &InputString,
 
   // Convert cmake command to lower case in cmake script files
   LineEndingString = convertCmakeCommandsToLower(LineEndingString);
+
+  std::map<std::string, std::string> VariablesMap;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/SaveNewFiles.cpp]: please put all the logic into MigrateCmakeScript.cpp and expose an interface.

File: clang/test/dpct/cmake_migation/case_009/rules.yaml
Comment (tomflinda) [reply clang/test/dpct/cmake_migation/case_009/rules.yaml:1]: ok

File: clang/lib/DPCT/SaveNewFiles.cpp (hunk 2 above)
Comment (tomflinda) [reply clang/lib/DPCT/SaveNewFiles.cpp]: Okay

File: clang/test/dpct/cmake_migation/case_009/expected.txt (hunk 1 above)
Comment (tomflinda) [reply clang/test/dpct/cmake_migation/case_009/expected.txt:1]: If each line in  input.cmake is checked by using "// CHECK:...", the checkline format like "// CHECK:..." in inpute.cmake will be treated as cmake script code.  So I suggest using current check way.

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 3]
```diff
@@ -200,3 +321,279 @@ void migrateCmakeScriptOnly(
...
+    const std::map<std::string, std::string> &VariablesMap) {
+  size_t Pos = Arg.find("...");
+  std::string ReplArg;
+  if (Pos != std::string::npos) {
+    const auto StrArray = split(Arg, "...");
+
+    std::string MinVer = getVarName(StrArray[0], VariablesMap);
+    std::string MaxVer = getVarName(StrArray[1], VariablesMap);
+
+    if (std::atof(MinVer.c_str()) >= 3.24) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp:468]: Please fix the magic number in follow up PR with variable or Macro.

File: clang/test/dpct/cmake_migation/case_009/input.cmake [hunk 4]
```diff
@@ -0,0 +1,10 @@
+CMake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.26)
+cmake_minimum_required(VERSION 3.10...3.26)
+set  (  VER 
+  3.15)
+set(MIN_VER 3.10)
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migation/case_009/input.cmake:6]: Please also consider add test case:  put set(MIN_VER 3.10) into separate file.  Please address it in a follow up PR.

**Review by zhimingwang36 — APPROVED** [review]
LGTM  ( Please address the comment in follow up PR).

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 5]
```diff
@@ -128,6 +218,37 @@ void collectCmakeScripts(StringRef InRoot, StringRef OutRoot,
...
+  std::string LineEndingString;
+  // pattern_rewriter require the input file to be LF
+  bool IsCRLF = fixLineEndings(InputString, LineEndingString);
+
+  // Convert cmake command to lower case in cmake script files
+  LineEndingString = convertCmakeCommandsToLower(LineEndingString);
+
+  std::map<std::string, std::string> VariablesMap;
+  parseVariable(LineEndingString, VariablesMap);
+  cmakeSyntaxProcessed(LineEndingString, VariablesMap);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp:234]: Please add comment on: Line 234 and L236.   what happen if both cmakeSyntaxProcessed and applyPatternRewriter process on same cmake syntax.

File: clang/test/dpct/cmake_migation/case_009/input.cmake (hunk 4 above)
Comment (tomflinda) [reply clang/test/dpct/cmake_migation/case_009/input.cmake:6]: okay

File: clang/lib/DPCT/MigrateCmakeScript.cpp (hunk 3 above)
Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp:468]: ok

**Pull request merged**